### ISSUE TITLE
:Man command is broken on OS X 10.10

### ISSUE THREAD

**Issue by pavoljuhas** [vim/vim#2488]
[This is to continue discussion in the closed issue #1956]

I found the following problem when using `:Man` command in the latest Vim.

```sh
$ vim -u NONE -i NONE -N -c 'source $VIMRUNTIME/ftplugin/man.vim' -c 'Man cp'

env: illegal option -- u
usage: env [-i] [name=value ...] [utility [argument ...]]
```

This happens because `/usr/bin/env` on OS X 10.10 does not accept the `-u` option.  The `env -u` was introduced in [7f2e9d7c](
https://github.com/vim/vim/commit/7f2e9d7c9cdfc5201a899b7b610edf64bf80c45f#diff-8691c83194ea5f1342ecc9f17b4c51d8L179) which I believe applied fix for #1956.  The issue considered a code using the `unset` shell builtin, but that would not work for the fish shell [[1956-comment]](https://github.com/vim/vim/issues/1956#issuecomment-343072112). 

I suggest to use `env MANPAGER=cat` instead, which should work across all shells and env versions.  As far as I can test, `env MANPAGER=cat man TOPIC` produces the same output as `(unset MANPAGER; man TOPIC)`.

@goweol - can you consider the following patch?
[0001-Fix-Man-command-on-Mac-OS-X-10.10.patch](https://github.com/vim/vim/files/1582993/0001-Fix-Man-command-on-Mac-OS-X-10.10.patch.txt)

```patch
From 01213be7bac601feea196e398062bc3cd6d376c5 Mon Sep 17 00:00:00 2001
From: Pavol Juhas <pavol.juhas@gmail.com>
Date: Fri, 22 Dec 2017 10:42:46 -0500
Subject: Fix `:Man` command on Mac OS X 10.10

Do not use non-portable `env -u MANPAGER`.
Set `MANPAGER=cat` instead.
---
 runtime/ftplugin/man.vim | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/runtime/ftplugin/man.vim b/runtime/ftplugin/man.vim
index cfcabc3fb..f6d2e370a 100644
--- a/runtime/ftplugin/man.vim
+++ b/runtime/ftplugin/man.vim
@@ -173,7 +173,7 @@ func <SID>GetPage(...)
 
   " Ensure Vim is not recursively invoked (man-db does this) when doing ctrl-[
   " on a man page reference by unsetting MANPAGER.
-  silent exec "r !env -u MANPAGER man ".s:GetCmdArg(sect, page)." | col -b"
+  silent exec "r !env MANPAGER=cat man ".s:GetCmdArg(sect, page)." | col -b"
 
   if unsetwidth
     let $MANWIDTH = ''
-- 
2.15.1

```

**Comment by goweol** [comment]
I'm fine with a patch if there's no regression.
Because I only uses ubuntu, I cannot test such things.

How about this patch?
It reverts 7f2e9d7c for man.vim and set MANPAGER to cat.

```
diff --git a/runtime/ftplugin/man.vim b/runtime/ftplugin/man.vim
index cfcabc3fb..f624e6311 100644
--- a/runtime/ftplugin/man.vim
+++ b/runtime/ftplugin/man.vim
@@ -1,7 +1,7 @@
 " Vim filetype plugin file
 " Language:	man
 " Maintainer:	SungHyun Nam <goweol@gmail.com>
-" Last Change: 	2017 Nov 11
+" Last Change: 	2017 Dec 27
 
 " To make the ":Man" command available before editing a manual page, source
 " this script from your startup vimrc file.
@@ -15,6 +15,14 @@ if &filetype == "man"
   endif
   let b:did_ftplugin = 1
 
+  " Ensure Vim is not recursively invoked (man-db does this)
+  " when doing ctrl-[ on a man page reference.
+  " There's no portable way to unset MANPAGER environment variable, so that
+  " let's just assign cat.
+  if exists("$MANPAGER")
+    let $MANPAGER = "cat"
+  endif
+
   " allow dot and dash in manual page name.
   setlocal iskeyword+=\.,-
 
@@ -171,9 +179,7 @@ func <SID>GetPage(...)
     let unsetwidth = 1
   endif
 
-  " Ensure Vim is not recursively invoked (man-db does this) when doing ctrl-[
-  " on a man page reference by unsetting MANPAGER.
-  silent exec "r !env -u MANPAGER man ".s:GetCmdArg(sect, page)." | col -b"
+  silent exec "r !man ".s:GetCmdArg(sect, page)." | col -b"
 
   if unsetwidth
     let $MANWIDTH = ''

```

**Comment by vim-ml** [comment]
2017-12-27 오후 1:52에 goweol (Vim Github Repository) 이(가) 쓴 글:
> I'm fine with a patch if there's no regression.
> Because I only uses ubuntu, I cannot test such things.
> 
> How about this patch?

Oops!  Sorry about annoying patch.
I should have check preview page. :-(
I updated patch in the github. Please check the patch there.

**Comment by goweol** [comment]
I think if statement is meaningless in the previous patch.
Here is a patch v2 which removes it.

```
diff --git a/runtime/ftplugin/man.vim b/runtime/ftplugin/man.vim
index cfcabc3fb..c20113030 100644
--- a/runtime/ftplugin/man.vim
+++ b/runtime/ftplugin/man.vim
@@ -1,7 +1,7 @@
 " Vim filetype plugin file
 " Language:	man
 " Maintainer:	SungHyun Nam <goweol@gmail.com>
-" Last Change: 	2017 Nov 11
+" Last Change: 	2017 Dec 27
 
 " To make the ":Man" command available before editing a manual page, source
 " this script from your startup vimrc file.
@@ -15,6 +15,12 @@ if &filetype == "man"
   endif
   let b:did_ftplugin = 1
 
+  " Ensure Vim is not recursively invoked (man-db does this)
+  " when doing ctrl-[ on a man page reference.
+  " There's no portable way to unset MANPAGER environment variable, so that
+  " let's just assign cat.
+  let $MANPAGER = "cat"
+
   " allow dot and dash in manual page name.
   setlocal iskeyword+=\.,-
 
@@ -171,9 +177,7 @@ func <SID>GetPage(...)
     let unsetwidth = 1
   endif
 
-  " Ensure Vim is not recursively invoked (man-db does this) when doing ctrl-[
-  " on a man page reference by unsetting MANPAGER.
-  silent exec "r !env -u MANPAGER man ".s:GetCmdArg(sect, page)." | col -b"
+  silent exec "r !man ".s:GetCmdArg(sect, page)." | col -b"
 
   if unsetwidth
     let $MANWIDTH = ''
```

**Comment by pavoljuhas** [comment]
> let $MANPAGER = "cat"

This would break the `K` command in Vim and also have an unexpected effect on the shell started with the `:shell` command.  I feel the version with `env MANPAGER=cat man ...` is better, because `MANPAGER=cat` is then in effect only for the `man` command that reads the text.  Otherwise MANPAGER remains unset or at user's preferred value set in the vim parent shell.

**Comment by goweol** [comment]
OK! I will send a patch which uses 'env MANPAGER=cat ...'.

**Comment by nuko8** [comment]
Hmm, but the change 'env MANPAGER=cat ...'. prevents `:Man` from working for me on OS X 10.11 or later due to busy loop.

**Comment by nuko8** [comment]
If we don't mind adding a special `if` branch dedicated to Mac to absorb the version difference, then the `if` statement shown below might be usable to address the issue:
```viml
if has('mac')
  let s:sw_vers_reply = systemlist('sw_vers')
  let s:product_version = substitute(s:sw_vers_reply[1], '\v\w+:\s+(\d+\.\d+\.\d+)', '\1', '')
  let s:minor_version = substitute(s:product_version, '\v\d+\.(\d+)\.\d+', '\1', '')

  if s:minor_version >= 11 " El Capitan or later
        " Use the current code.
  else " Yosemite or earlier
    " Use the proposed 'env MANPAGER=cat ...' code.
  endif
else
  " Use the current code (or the code that had been used prior to the latest
  " man.vim update).
endif
```
Definitely the code above doesn't look cool at all, but hopefully, it would be at least helpful to make the idea clear...

**Comment by pavoljuhas** [comment]
>  'env MANPAGER=cat ...'. prevents :Man from working for me on OS X 10.11 or later due to busy loop.

I don't understand why should this happen.
Do you see the same problem with `man` run from the shell after `export MANPAGER=cat`?
Do you get that issue in a vim session started as
```sh
$ vim -u NONE -i NONE -N -c 'source $VIMRUNTIME/ftplugin/man.vim'
```
Are you using translated man pages?

**Comment by nuko8** [comment]
> Do you see the same problem with man run from the shell after export MANPAGER=cat?

No.

> Do you get that issue in a vim session started as
> 
> $ vim -u NONE -i NONE -N -c 'source $VIMRUNTIME/ftplugin/man.vim'

No.

> Are you using translated man pages?

No.

> I don't understand why should this happen.

As I explained somewhere, the previous man.vim implementation set MANPAGER empty all the way to avoid recursive calls.  Your proposal obviously breaks it; that breakage is needed to be addressed before your patch is accepted.

**Comment by pavoljuhas** [comment]
> > Do you get that issue in a vim session started as
> > $ vim -u NONE -i NONE -N -c 'source $VIMRUNTIME/ftplugin/man.vim'

> No.

Assuming `$VIMRUNTIME/ftplugin/man.vim` is the version with `env MANPAGER=cat` the problem has to be due to some other interfering plugin or setting in your vimrc.  Perhaps this should be tracked down and understood before claiming MANPAGER=cat creates a problem.

Can you post a reproducible example of getting into the busy loop starting from `vim -u NONE -i NONE -N`?

**Comment by nuko8** [comment]
> Assuming $VIMRUNTIME/ftplugin/man.vim is the version with env MANPAGER=cat the problem has to be due to some other interfering plugin or setting in your vimrc. 

My plugins for daily use are the default ones, a few of tpope's, and two or three simple ones which I made myself.  I can't see how they are interfering with the patched man.vim.

Also, how can my vimrc ~is~ be relevant?  True, it has a few thousands of lines, but I carefully followed your instructions and did:

>> Do you get that issue in a vim session started as
>> $ vim -u NONE -i NONE -N -c 'source $VIMRUNTIME/ftplugin/man.vim'

Obviously, .vimrc ~shouldn't~ should be irrelevant because of `-u NONE`.

> Can you post a reproducible example of getting into the busy loop starting from vim -u NONE -i NONE -N?

Run Vim on OS X 10.11 or later and issue `:Man` for an arbitrary man page with **a man.vim modified with your patch**.

**Comment by pavoljuhas** [comment]
> My plugins for daily use are the default ones, a few of tpope's, and two or three simple ones which I made myself. I can't see how they are interfering with the patched man.vim.

@nuko8 - the way I understand [your earlier comment](https://github.com/vim/vim/issues/2488#issuecomment-354128401) is that `:Man` works as expected when you run vim using
```shell
$ vim -u NONE -i NONE -N -c 'source /path/to/patched/ftplugin/man.vim'
```
i.e., it works when you *avoid* loading your .vimrc and plugins.  
So does `:Man` work in such session or no?

If it works there, but not in a regular `vim`, it means that either your plugins or .vimrc do something that makes the `:Man` command fail.  Also, please check the `:scriptnames` output to make sure vim loads the correct version of `ftplugin/man.vim`.

> Run Vim on OS X 10.11 or later and issue :Man for an arbitrary man page with a man.vim ...

Unfortunately, I don't have access to a machine with OS X 10.11. 

What is the MANPAGER value in your shell?
If you set it to something, what startup file do you use to do so?

For me (OS X 10.10) the patch works well with `unset MANPAGER` and also with `export MANPAGER="env MAN_PN=1 vim -M +MANPAGER -"`.

**Comment by nuko8** [comment]
@pavoljuhas

I believe I already fully answered your questions.  Let's stop repeating similar things and step forward to fixing the issue.

@goweol 

The following patch should make man.vim support macOS <= 10.10 as well as recent versions:
```viml
diff --git a/man.vim.orig b/man.vim
index cfcabc3..3af9eef 100644
--- a/man.vim.orig
+++ b/man.vim
@@ -171,9 +171,18 @@ func <SID>GetPage(...)
     let unsetwidth = 1
   endif
 
+  let s:man_cmd = "!env -u MANPAGER man "
+  if has('mac')
+    let s:osx_min_ver = substitute(system('sw_vers -productVersion'), '\v\d+\.(\d+)(\.\d+)?', '\1', '')
+    if str2nr(s:osx_min_ver) < 11
+      " env(1) of 10.10 (Yosemite) or earlier doesn't accept -u.  Use cat(1)
+      " instead.
+      let s:man_cmd = "!env MANPAGER=cat man "
+    endif
+  endif
   " Ensure Vim is not recursively invoked (man-db does this) when doing ctrl-[
   " on a man page reference by unsetting MANPAGER.
-  silent exec "r !env -u MANPAGER man ".s:GetCmdArg(sect, page)." | col -b"
+  silent exec "r ".s:man_cmd.s:GetCmdArg(sect, page)." | col -b"
 
   if unsetwidth
     let $MANWIDTH = ''
```

**Comment by goweol** [comment]
@nuko8 

I have no objection to your patch.
But I don't know if it would be a good idea to do so.
Acording to Christian's comment, 'env -u' is not portable. All the other system supports it?
We know mac osx 10.10 does not support 'env -u', does it true for older mac?
If VIM supports such a 'unsetenv' command, can't we use it for this situation?  I know we have to supports older VIM, but for future.

After all the dicussion is settled down and all the man.vim users have no problem with this patch, it will be applied.

Thanks!

**Comment by pavoljuhas** [comment]
Indeed, [that patch](https://github.com/vim/vim/issues/2488#issuecomment-354225162) would fail on non-mac systems that do not support `env -u`.

Here is a different take that unsets MANPAGER by explicitly running `/bin/sh`.  That should work regardless of the 'shell' option value and will use the same man-reading command on all systems.

[0001-Avoid-non-portable-use-of-env-in-the-Man-command.patch.txt](https://github.com/vim/vim/files/1592484/0001-Avoid-non-portable-use-of-env-in-the-Man-command.patch.txt)

```patch
diff --git a/runtime/ftplugin/man.vim b/runtime/ftplugin/man.vim
index cfcabc3fb..3bf2c2932 100644
--- a/runtime/ftplugin/man.vim
+++ b/runtime/ftplugin/man.vim
@@ -173,7 +173,11 @@ func <SID>GetPage(...)
 
   " Ensure Vim is not recursively invoked (man-db does this) when doing ctrl-[
   " on a man page reference by unsetting MANPAGER.
-  silent exec "r !env -u MANPAGER man ".s:GetCmdArg(sect, page)." | col -b"
+  " Because Vim 'shell' may be set to the fish shell, we run /bin/sh
+  " explicitly to make sure that the `unset` command is available.
+  let man_cmd = '/bin/sh -c "' .
+	\ 'unset MANPAGER; man ' . s:GetCmdArg(sect, page) . '" | col -b'
+  silent exec "r !" . man_cmd
 
   if unsetwidth
     let $MANWIDTH = ''
```

> If VIM supports such a 'unsetenv' command, can't we use it for this situation?

AFAIK, there is no way to unset an environment variable with a Vim script yet. #1116

**Comment by nuko8** [comment]
@goweol 

> But I don't know if it would be a good idea to do so.
> Acording to Christian's comment, 'env -u' is not portable. All the other system supports it?

Good point and also one of my concerns about the issue; however, I made the patch based on the following my thoughts:  

1. It is true that `env -u` is not portable.  That said, none of us had noticed it and heard someone complaining about it until the fact that it doesn't work for *older* Mac's was reported here.   So, other than that exceptional case, it's natural to *assume* for now that man.vim does work for the rest.  Accordingly, keeping the existing code path should make sense in avoiding or minimizing any possible regression which may happen for systems for which man.vim is already working well.

2. As to portability, it's clear that none of us can test the proposed patches with every platform where man.vim is supposed to work.  In fact, we've been only discussing the portability issue in quite an abstracted manner by referring to some standards or specifications.  For such cases, I think it better to wait for someone's concrete report on portability (and then to try fixing issues raised by such reports), and that, rather than spending time debating possible hypothetical future issues, we should try fixing the actual, concrete issue we are facing now.   The construction of my patch is preparing for the day when someone reports that man.vim doesn't work for her/his environment; all we need to do then is to set up a new `man_cmd` tailored for her/him by adding a new `else if` branch (and I expect that the number of it will be two or three, if any).

> We know mac osx 10.10 does not support 'env -u', does it true for older mac?

While the man page of `env` of 10.11 or later is dated 2008, that of 10.10 was 1993 or so.  Since (desktop) Mac OS  X was launched in 2001, I suppose that all the older Mac OS X's prior to 10.10 ~has~ have the same `env` as that of 10.10.

> If VIM supports such a 'unsetenv' command, can't we use it for this situation? I know we have to supports older VIM, but for future.

Ah, that's another portability issue, as Bram commented in #1116 which @pavoljuhas already referred to.   That would definitely make life easier if we had a portable one...

**Comment by goweol** [comment]
@nuko8 

Your patch checks only minor version.  It will not work if apple release osx 11.0.
Is it OK?

@pavoljuhas 

I will take nuko8's patch because I agree with his thought about portability.

Thanks,

**Comment by pavoljuhas** [comment]
> I will take nuko8's patch because I agree with his thought about portability.

@goweol - I don't think that patch is a good solution.  First, it does not fix the issue on non-Mac systems with an older version of env.  OS X commands derive from BSD so Vim on an older BSD versions would still have `:Man` failing on `env -u`.  Second, it adds unnecessary system-dependent `if` branch which makes the code more complicated, more difficult to follow and thus trickier to maintain in the future.  In my experience simpler is better.

My [second patch](https://github.com/vim/vim/issues/2488#issuecomment-354390150) does in effect the same as `env -u`, but uses `/bin/sh` instead.  It is to my knowledge portable to all UNIX-like systems, and does not introduce extra branching.  I can't foresee a scenario where it fails - if you know of some, please let me know.

If you still decide to go for nuko8's approach I'd strongly recommend the following improvements:

1. use a direct test whether `env` accepts `-u` instead of a fragile parsing of OS version (what if user installs updated `env` from macports?)
2. cache the result in a script variable, say `s:env_has_u` to avoid an extra (expensive) `system()` call per each `:Man` command

Example:

```VimL
  if !exists('s:env_has_u')
    call system('env -u x true')
    let s:env_has_u = (v:shell_error == 0)
  endif
  let man_cmd = s:env_has_u ? 'env -u MANPAGER man ' : 'env MANPAGER=cat man '
  ...
```

**Comment by nuko8** [comment]
@goweol 

> Your patch checks only minor version. It will not work if apple release osx 11.0.  Is it OK?

To my knowledge, the X of Mac OS X represents the Roman numeral meaning 10 (ten),  so I couldn't think about the case where the version number ever got 11.0 or larger for that OS.   But it is now called macOS and the X is dropped.  Probably, revising the patch and making it future-proof could be a good idea:
```diff
diff --git a/man.vim.orig b/man.vim
index cfcabc3..48a9ddc 100644
--- a/man.vim.orig
+++ b/man.vim
@@ -171,9 +171,20 @@ func <SID>GetPage(...)
     let unsetwidth = 1
   endif
 
+  let s:man_cmd = "!env -u MANPAGER man "
+  if has('mac')
+    let s:macos_maj_ver = substitute(system('sw_vers -productVersion'), '\v(\d+)\.\d+(\.\d+)?', '\1', '')
+    let s:macos_min_ver = substitute(system('sw_vers -productVersion'), '\v\d+\.(\d+)(\.\d+)?', '\1', '')
+    if str2nr(s:macos_maj_ver) == 10 && str2nr(s:macos_min_ver) < 11
+      " env(1) of 10.10 (Yosemite) or earlier doesn't accept -u.  Use cat(1)
+      " instead.
+      let s:man_cmd = "!env MANPAGER=cat man "
+    endif
+    echom s:macos_maj_ver . ' ' . s:macos_min_ver
+  endif
   " Ensure Vim is not recursively invoked (man-db does this) when doing ctrl-[
   " on a man page reference by unsetting MANPAGER.
-  silent exec "r !env -u MANPAGER man ".s:GetCmdArg(sect, page)." | col -b"
+  silent exec "r ".s:man_cmd.s:GetCmdArg(sect, page)." | col -b"
 
   if unsetwidth
     let $MANWIDTH = ''
```

@pavoljuhas 

Let's reconfirm the goal of the issue we've been discussing so far: Resolve ":Man command is broken on OS X 10.10".  Let's fix it first and get us ready for other possible issues in the future...

**Comment by nuko8** [comment]
Oops, the patch I've just sent includes unnecessary code for debugging (= A line beginning with `echom`).  Please use the following instead:
```diff
diff --git a/man.vim.orig b/man.vim
index cfcabc3..fe70035 100644
--- a/man.vim.orig
+++ b/man.vim
@@ -171,9 +171,19 @@ func <SID>GetPage(...)
     let unsetwidth = 1
   endif
 
+  let s:man_cmd = "!env -u MANPAGER man "
+  if has('mac')
+    let s:macos_maj_ver = substitute(system('sw_vers -productVersion'), '\v(\d+)\.\d+(\.\d+)?', '\1', '')
+    let s:macos_min_ver = substitute(system('sw_vers -productVersion'), '\v\d+\.(\d+)(\.\d+)?', '\1', '')
+    if str2nr(s:macos_maj_ver) == 10 && str2nr(s:macos_min_ver) < 11
+      " env(1) of 10.10 (Yosemite) or earlier doesn't accept -u.  Use cat(1)
+      " instead.
+      let s:man_cmd = "!env MANPAGER=cat man "
+    endif
+  endif
   " Ensure Vim is not recursively invoked (man-db does this) when doing ctrl-[
   " on a man page reference by unsetting MANPAGER.
-  silent exec "r !env -u MANPAGER man ".s:GetCmdArg(sect, page)." | col -b"
+  silent exec "r ".s:man_cmd.s:GetCmdArg(sect, page)." | col -b"
 
   if unsetwidth
     let $MANWIDTH = ''
```

**Comment by pavoljuhas** [comment]
> Let's reconfirm the goal of the issue we've been discussing so far: Resolve ":Man command is broken on OS X 10.10". Let's fix it first and get us ready for other possible issues in the future...

@nuko8 - I have identified the problem on OS X 10.10, but it certainly happens on all other UNIX-derived systems where `env` does not take the `-u` option.  Vim runs on all UNIX platforms, and it is therefore better to fix :Man for them all.  Your patch plugs the hole on OS X, but leaves it open everywhere else.  My version fixes the problem on all platforms and happens to be shorter and simpler.  I fail to see a technical reason for choosing your version over mine.

cc: @goweol

**Comment by mattn** [comment]
Please do not use `/bin/sh` directory. Current implementation of man.vim is working well on Windows.

**Comment by pavoljuhas** [comment]
> Please do not use /bin/sh directory. Current implementation of man.vim is working well on Windows.

I see, maybe it is better to use `env` after all.  It is a smaller command wrapper than `sh`.  Here is a patch similar to the nuko8 version, but with a generic test for the `-u` support.  

@mattn - will this work for you?  I assume if your Windows has `env` and `man` it also has the `true` command.

```diff
diff --git a/runtime/ftplugin/man.vim b/runtime/ftplugin/man.vim
index cfcabc3fb..fb686e55c 100644
--- a/runtime/ftplugin/man.vim
+++ b/runtime/ftplugin/man.vim
@@ -173,7 +173,15 @@ func <SID>GetPage(...)
 
   " Ensure Vim is not recursively invoked (man-db does this) when doing ctrl-[
   " on a man page reference by unsetting MANPAGER.
-  silent exec "r !env -u MANPAGER man ".s:GetCmdArg(sect, page)." | col -b"
+  " Some versions of env(1) do not support the '-u' option, and in such case
+  " we set MANPAGER=cat.
+  if !exists('s:env_has_u')
+    call system('env -u x true')
+    let s:env_has_u = (v:shell_error == 0)
+  endif
+  let env_cmd = s:env_has_u ? 'env -u MANPAGER' : 'env MANPAGER=cat'
+  let man_cmd = env_cmd . ' man ' . s:GetCmdArg(sect, page) . ' | col -b'
+  silent exec "r !" . man_cmd
 
   if unsetwidth
     let $MANWIDTH = ''
```

[0001-Avoid-use-of-env-u-when-not-supported.patch](https://github.com/vim/vim/files/1605444/0001-Avoid-use-of-env-u-when-not-supported.patch.txt)

**Comment by jamessan** [comment]
> Please do not use /bin/sh directory. Current implementation of man.vim is working well on Windows.

In what configuration does Windows have the man command?  If it has man, why wouldn't it also have sh or env?  Is the concern solely about using `/bin/sh` instead of `sh`?

**Comment by mattn** [comment]
msys2 mingw64.  current implementation doesn't use sh nor /bin/sh.

**Comment by chrisbra** [comment]
> msys2 mingw64. current implementation doesn't use sh nor /bin/sh.

But it has one, right?

Agree with @pavoljuhas  latest [patch](https://github.com/vim/vim/issues/2488#issuecomment-355464096)

**Comment by mattn** [comment]
I also +1 to the patch since it doesn't use sh or /bin/sh.

**Comment by goweol** [comment]
If man.vim needs to care about the MANPAGER variable, I will inlcude [@pavoljuhas's latest patch](https://github.com/vim/vim/issues/2488#issuecomment-355464096.)

I don't remember man.vim was affected by manpager.vim or MANPAGER environment variable (not related to manpager.vim).  Because @Konfekt suggests new manpager.vim which does not use man.vim, I guess I can safely remove the MANPAGER stuff from man.vim with his new manpager.vim.

**Comment by nuko8** [comment]
> Because @Konfekt suggests new manpager.vim which does not use man.vim, I guess I can safely remove the MANPAGER stuff from man.vim with his new manpager.vim.

FYI, it's true that the new manpager.vim stops directly invoking `:Man` from the plugin, but it sources `ftplugin/man.vim` instead and still relies on the value of MANPAGER...

**Comment by nuko8** [comment]
> it sources ftplugin/man.vim instead

The relevant code is: https://github.com/vim/vim/pull/2529/files#diff-049b3b4a0ca427766767f52897d4af7cR15

**Comment by pavoljuhas** [comment]
> Because @Konfekt suggests new manpager.vim which does not use man.vim, I guess I can safely remove the MANPAGER stuff from man.vim with his new manpager.vim.

As long as your plugin uses `man` to get the content, it is safer to run it with unset MANPAGER. MANPAGER could be set to anything in the invoking shell, e.g., MANPAGER=head.  The `MANPAGER="env MAN_PN=1 vim -M +MANPAGER -"` suggested in Vim help would make `:Man` start another Vim instance.

**Comment by pavoljuhas** [comment]
Fixed by 40962ec9c0e7b8699e101182b06ddd39dc0e1212.
